### PR TITLE
requests redux modules

### PR DESCRIPTION
**Pull request by muselesscreator**
Adds a new `requests` redux reducer which tracks the state of a specified set of network requests.

Request identifiers (RequestKeys) and states (RequestStates) are specified in StrictDict keystores in `src/data/constants`

New `thunkActions/requests` module includes a `networkRequest` thunkAction that takes a key, a network event promise, and optional onSuccess and onFail callbacks, and triggers setRequestState action on fire, completion, or failure.

All API calls are wrapped around this with tiny wrapper functions before consumption in their relevant thunkActions files.

All the stuff I touched is also unit tested now.

*Bonus points* a bug got introduced with how ungraded-but-locked-by-you-from-another-session had a null (vs undefined) gradeData entry, which messed with the check for whether an empty grade should be generated.  Now just checks truthiness.

JIRA: [AU-338](https://openedx.atlassian.net/browse/AU-338)

image

### CODE REVIEW

**Review by mattcarter — APPROVED** [review]
Approved

File: src/data/thunkActions/requests.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ */
+export const networkRequest = ({
+  requestKey,
+  promise,
+  onSuccess,
+  onFailure,
+}) => (dispatch) => {
+  dispatch(actions.requests.startRequest(requestKey));
+  return promise.then((response) => {
+    dispatch(actions.requests.completeRequest({ requestKey, response }));
```
Comment (muselesscreator) [inline src/data/thunkActions/requests.js]: Move logging to after callbacks

File: src/data/thunkActions/requests.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    requestKey: RequestKeys.initialize,
+    promise: api.initializeApp(locationId),
+    ...rest,
+  }));
+};
+
+/**
+ * Tracked fetchSubmissionResponse api method.
+ * Tracked either prefetchNext or prefetchPrev request key.
+ * @param {string} submissionId - target submission id
```
Comment (muselesscreator) [inline src/data/thunkActions/requests.js:53]: include requestKey param

**Review by jnlapierre — APPROVED** [review]
Walkthrough with Ben W and Leangseu, pulled PR and ran app and tests locally, LGTM

File: src/data/reducers/requests.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    [payload.requestKey]: {
+      status: RequestStates.completed,
+      response: payload.response,
+    },
+  }),
+  [actions.requests.failRequest]: (state, { payload }) => ({
+    ...state,
+    [payload.requestKey]: {
+      status: RequestStates.failed,
+      error: payload.error,
```
Comment (leangseu-edx) [inline src/data/reducers/requests.js:35]: resetting response on error? no?

Comment (muselesscreator) [reply src/data/reducers/requests.js:35]: yep.  clearing response when there is an error, as it would be associated with a previous message

**Pull request merged**